### ISSUE TITLE
Cargo make fails when run from an windows UNC mounted network drive 

### ISSUE THREAD

**Issue by virtual-hexagon** [sagiegurari/cargo-make#581]
### What I am trying to do
The cargo make file I am trying to execute is simply a Hello World from the powershell task runner. My project is located on a network drive with the UNC path `\\?\UNC\ULCV\projs\murray\Code\mango`.  The UNC path is mapped to the system as `P:\murray\Code\mango`.  For compatibility with tools I often have to execute a ` pushd . ` command before spawning any command line applications. See evanmcdonnal's accepted answer on this stack exchange for an explanation of why: [link](https://superuser.com/questions/282963/browse-an-unc-path-using-windows-cmd-without-mapping-it-to-a-network-drive)

### Describe The Bug

Cargo make fails in my environment when run from a UNC mounted network drive. Specifically it seems that I cannot control the current working directory that is passed to CMD.exe when it is spawned during the workspace task. I have attempted to override the UNC path by setting some environment variables but they do not seem to be respected. I suspect that injecting a `pushd .` before cmd.exe attempts to change directories or spawning the cmd.exe process would resolve this problem, but am too much of a rust newb to really get a good grip on where the cmd.exe process is being driven from within the project.

### To Reproduce

Clone a git project containing cargo make onto a windows network file share mounted under Network File System (NFS). The drive should appear with a drive letter, but is also addressable as a UNC network path.  Create a cargo make task that contains the powershell runner and have it output 'Hello World' to console. cargo make should fail during the workspace task when cmd.exe  is spawned.  cmd.exe does not understand UNC paths and thus cannot change directory to the location of the project.

### Things I have tried
I have attempted to set the environment var $CARGO_MAKE_WORKING_DIRECTORY to my project root that contains the makefile.toml as a non UNC path, but the variable did not seem to impact this failure that I am seeing. I have not attempted to set $CARGO_MAKE_WORKSPACE_WORKING_DIRECTORY yet but will try that soon.

I have also attempted to control the environment by performing a `pushd .` before executing the cargo make command, but that did not seem to help either. As a next step I will try moving my project to the `C:\` drive to test if cargo make will work with
 my project, unfortunately that may not be a long term solution for me due to limitations on my `C:\` drive.

### Related info
I found this issue on the GitHub Page: #561, in which  @WilliamVenner addressed some UNC path issues.  After updating to the latest version of cargo make, so my tool chain could include the fix from issue #561,  cargo make will get further along the build before failing but still has UNC pathing issues in my current environment.

### Code Sample
Relevant content of makefile.toml: 

```rust
[tasks.powershell]
script_runner = "powershell"
script_extension = "ps1"
script = '''
Write-Host "Hello, World!"
'''
```
Full File:
[mango_toml.zip](https://github.com/sagiegurari/cargo-make/files/7108506/mango_toml.zip)

### Error Stack

Note that CMD.exe falls over because it cannot interpret a UNC path. cargo make then falls back to C:\  as the default drive and attempts to PUSHD into my cargo workspace path: proj\frontend.  that folder does not exist at the C:\ drive location so the task fails.

```console
(base) PS P:\murray\Code\mango> cargo make --makefile .\mango.toml --task powershell
[cargo-make] INFO - cargo make 0.35.0
[cargo-make] INFO - Build File: .\mango.toml
[cargo-make] INFO - Task: powershell
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: workspace
'\\?\UNC\ULCV\projs\murray\Code\mango'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.

C:\Windows>cd "\\?\UNC\ULCV\projs\murray\Code\mango"
'\\?\UNC\ULCV\projs\murray\Code\mango'
CMD does not support UNC paths as current directories.

C:\Windows>PUSHD proj\frontend
The system cannot find the path specified.

C:\Windows>cargo make --disable-check-for-updates --allow-private --no-on-error --loglevel=info --env CARGO_MAKE_CRATE_CURRENT_WORKSPACE_MEMBER=frontend --profile development -- powershell
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: powershell
[cargo-make][1] INFO - Profile: development
[cargo-make][1] ERROR - Task powershell not found
[cargo-make][1] WARN - Build Failed.

C:\Windows>if 1 NEQ 0 exit /b 1
[cargo-make] ERROR - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

**Comment by sagiegurari** [comment]
thanks, i don't have a windows machine for dev at the moment, but i can try to do code review based on your detailed output and see if i can find something.

**Comment by virtual-hexagon** [comment]
I am more than happy to help with testing if you throw a patch or branch my way to run through my environment.

**Comment by sagiegurari** [comment]
@virtual-hexagon I hope i fixed it (no idea), can you test with the 0.35.1 branch please?

**Comment by virtual-hexagon** [comment]
@sagiegurari I built the 0.35.1 branch from source  on windows and ran it in my environment. Unfortunately it did not appear to fix the problem and the behavior of the error is still unchanged. I am reviewing my verbose logs a little closer to see if I spot any changes in behavior. 

I noticed that you pushed the dunce dependency down into the run_script crate, so I plan to review the run_script code a bit closer. perhaps I can log some output / run through with the debugger to better understand what is happening there when it is handling the failing task.

**Comment by sagiegurari** [comment]
@virtual-hexagon i saw the 'workspace' task fails for you.
It is a special task that is auto generated as a script and executed which in turn invokes each member.
I've migrated that native script to duckscript which runs in proc so workspace task should no longer fail for you.
i'm not sure if this will resolve the entire issue since more scripts may be invoked from your makefiles and i want to know if that will work or not. so would love you get some feedback on latest changes when you can.

**Comment by virtual-hexagon** [comment]
@sagiegurari  I pulled down the latest update. Good news, I was able to get my project to hello world from powershell. But I ran across some behavior that I am not sure if it is intended or not.

Originally when I executed the updated cargo-make, I ran across an error. After looking at verbose logs, I determined that the new workspace script could not find a task called 'powershell' because it was searching for a Makefile.toml that did not exist in my project. The generated script was searching inside of the workspace dir's and not at the root of the entire project where the command was executed from.

``` Powershell
(base) PS P:\murray\Code\mango> cargo make --makefile .\mango.toml powershell
[cargo-make] INFO - cargo make 0.35.1
[cargo-make] INFO - Build File: .\mango.toml
[cargo-make] INFO - Task: powershell
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: workspace
\\?\UNC\ULCV\projs\murray\Code\mango
[cargo-make][1] INFO - Project: mango
[cargo-make][1] INFO - Build File: Makefile.toml    (  < -- note this line  )
[cargo-make][1] INFO - Task: powershell             (  < -- note this line  )
[cargo-make][1] INFO - Profile: development
[cargo-make][1] ERROR - Task powershell not found
[cargo-make][1] WARN - Build Failed.
[cargo-make] ERROR - Error while running duckscript: Source: Unknown Line: 4 - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

To get my project to work I had to copy the Makefile into the root of each workspace and have a task called powershell defined in each file. I am unsure if this is intended behavior or by copying the files I am just tricking the duckscript into working.

Two things strike me as odd:

1) What is the point of specifying the top level Makefile at the root of the entire project if I then need to the have duplicates at the root of each workspace dir. This seems redundant and does not line up with how I interpreted the documentation.

2) When cargo make was searching for the Makefile in the root of each workspace directory, It was searching for the default name of 'Makefile.toml.'  Eventually I would like to name the make file in the frontend and backend workspaces something appropriate rather than a generic name like Makefile.toml.  This may already be possible and I just need to read the docs again.

``` Powershell
.    ( <-- cargo make command command executed here )
| - mango.toml
| - proj
|    | - frontend
|        | - Makefile.toml  ( <-- copy of mango.toml )
|    | - backend
|        | - Makefile.toml  ( <-- copy of mango.toml ) 
```


With this change I now I get the expected behavior :

``` Powershell
(base) PS P:\murray\Code\mango> cargo make --makefile .\mango.toml powershell
[cargo-make] INFO - cargo make 0.35.1
[cargo-make] INFO - Build File: .\mango.toml
[cargo-make] INFO - Task: powershell
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: workspace
\\?\UNC\ULCV\projs\murray\Code\mango
[cargo-make][1] INFO - Project: mango
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: powershell
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: powershell
[cargo-make][1] INFO - Execute Command: "powershell" "C:\\Users\\wagha\\AppData\\Local\\Temp\\fsio\\1N79txaAvX.ps1"
Hello, World!
[cargo-make][1] INFO - Build Done in 52.02 seconds.
[cargo-make][1] INFO - Project: mango-server
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: powershell
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Running Task: powershell
[cargo-make][1] INFO - Execute Command: "powershell" "C:\\Users\\wagha\\AppData\\Local\\Temp\\fsio\\Px7VtvquoK.ps1"
Hello, World!
[cargo-make][1] INFO - Build Done in 45.82 seconds.
[cargo-make] INFO - Build Done in 288.14 seconds.
```

**Comment by sagiegurari** [comment]
1. each directory is handled on it own. so you don't get root makefiles automatically.
this allows you some isolation to do things very differently, for example a project with javascript frontend code and rust backend code. in which build means something else for each one.
there is a way to extend the root makefile via one liner
https://github.com/sagiegurari/cargo-make#usage-workspace-extend
```
[env]
CARGO_MAKE_EXTEND_WORKSPACE_MAKEFILE = true
```

2. for standalone thats simple, just cli param --makefile.
however, for workspaces, I do not currently have that support.
basically it boils down to, how would cargo-make know that in one dir its called a.toml and in another dir its called b.toml
introducing some config file in those directories is not simpler that just doing this in the member dir

```toml
extend = "mango.toml"
```

and thats it.

**Comment by virtual-hexagon** [comment]
@sagiegurari okay I see what you are saying and that makes sense!  As I was messing around with the tasks & flows some more I  ran across a strange edge case. I can run my hello world powershell tasks fine, but if I run a cargo based command such as cargo clean then I get a new error while resolving workspaces.

```pwsh
(base) PS P:\murray\Code\mango> cargo make clean
[cargo-make] INFO - cargo make 0.35.1
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: clean
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Running Task: workspace
\\?\UNC\ULCV\projs\murray\Code\mango
[cargo-make][1] INFO - Project: mango
[cargo-make][1] INFO - Build File: Makefile.toml
[cargo-make][1] INFO - Task: clean
[cargo-make][1] INFO - Profile: development
[cargo-make][1] INFO - Skipping Task: legacy-migration
[cargo-make][1] INFO - Execute Command: "cargo" "clean"
error: current package believes it's in a workspace when it's not:
current:   \\?\UNC\ULCV\projs\murray\Code\mango\proj\frontend\Cargo.toml
workspace: \\?\UNC\ULCV\projs\murray\Code\mango\Cargo.toml

this may be fixable by adding `proj\frontend` to the `workspace.members` array of the manifest located at: \\?\UNC\ULCV\projs\murray\Code\mango\Cargo.toml
Alternatively, to keep it out of the workspace, add the package to the `workspace.exclude` array, or add an empty `[workspace]` table to the package's manifest.
[cargo-make][1] ERROR - Error while executing command, exit code: 101
[cargo-make][1] WARN - Build Failed.
[cargo-make] ERROR - Error while running duckscript: Source: Unknown Line: 4 - Error while executing command, exit code: 1
[cargo-make] WARN - Build Failed.
```

I can resolve this error by changing my workspace file as follows:
```rust
[workspace]                  [workspace]
members = [                  members = [
    "proj/frontend",    ->    "proj\\frontend",
    "proj/backend",     ->    "proj\\backend",
]                            ]
```

after making the above changes then the cargo based tasks will pass.

**Comment by sagiegurari** [comment]
@virtual-hexagon that feels like a cargo issue (not cargo-make) which fails to find the members

**Comment by sagiegurari** [comment]
@virtual-hexagon the original issue is now released in new official cargo-make release. thanks for the help testing.
if you have more issues please open them as new ones.
the current issue you are facing seems more like cargo issue than cargo-make.

**Comment by virtual-hexagon** [comment]
@sagiegurari  Yay,  thanks for all of the help!  

I agree that this problem sounds like cargo problem, and either way I have a path forward for my project thanks to your help.

**Comment by geovex** [comment]
I'm little lost. Is there any way to use cmd runner (or any command for that matter) without UNC path on mounted drive in windows. Because when i run any command i get something like this
```
PS Z:\project> cargo make build-win
[cargo-make] INFO - cargo make 0.36.0
[cargo-make] INFO - Project: project
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: build-win
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Execute Command: "7z"

Unhandled Exception: System.ArgumentException: Illegal characters in path.
   at System.Security.Permissions.FileIOPermission.EmulateFileIOPermissionChecks(String fullPath)
   at System.IO.Directory.InternalGetCurrentDirectory(Boolean checkHost)
   at shim.ShimProgram.Main(String[] args)
[cargo-make] ERROR - Error while executing command, exit code: -532462766
[cargo-make] WARN - Build Failed.
```

**Comment by sagiegurari** [comment]
can you run it with verbose logs so we can see more info? maybe put the output in a gist and share the link in your response so it would be more readable.

you can also replace the runner instead of cmd if it helps.
another workaround would be to use duckscript (not sure if it will help or not....) and run it from there.

to be honest, i don't have a windows development machine to fully support such issues and i need your help to resolve.

**Comment by geovex** [comment]
I tried to exclude side effects of other tools. And made a kind-of clean rust-msvc install, cargo-make and hello-world project and on mounted drive. But still
`Makefile.toml`
```toml
[tasks.build-win]
command = "cmd"
args = ["/C", "cd"]
```
<details>
<summary> short output </summary>

```
PS Z:\temp\hello> cargo make build-win
[cargo-make] INFO - cargo make 0.36.0
[cargo-make] INFO - Project: hello
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: build-win
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] INFO - Execute Command: "cmd" "/C" "cd"
'\\?\UNC\m25-600-10184\work$\temp\hello'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
C:\Windows
[cargo-make] INFO - Build Done in 1.04 seconds.
```
</details>

also with `-l verbose`
[verbose.log](https://github.com/sagiegurari/cargo-make/files/9469697/verbose.log)

Is this an expected behaviour or is there a way to supply drive:path as current working dir?

**Comment by sagiegurari** [comment]
@geovex thanks. please open a new issue with this info and i'll try to help.
i do'nt have windows, but maybe with the log you gave, i can try to help.
i'll probably need your help to test things out
